proposals/p2240.md

```diff
@@ -0,0 +1,689 @@
+element of that argument. The details of the rewrite vary slightly depending on
+the root node of the expansion:
+
+-   In a `...{` expansion, each instance uses a `{` in place of the opening
+    `...{`.
+-   In a `...,`, `...and`, or `...or` expansion, each instance has the expansion
+    operator removed, and the instances are joined using the corresponding
+    non-expansion operator.
+
+An `...and` expansion with arity 0 is rewritten to `true` and an `...or`
+expansion with arity 0 is rewritten to `false` (the identities of the respective
+logical operations). A `...,` expansion with arity 0 is rewritten to `,`, and
+any `,` immediately before or after the expansion is removed.
+
+Even though an `[:]` operator can only appear within a pack expansion, it is
+possible to apply these rewrite semantics to an isolated subexpression that
+contains a `[:]` operator but not the enclosing pack operator. In those cases,
+the result of the rewrite is an expression headed by a `[:]` operator, which
+means that its value is a pack. Thus, we can say that
+`([:]("a", "b", "c"), [:](1, 2, 3))` rewrites to
+`[:](("a", 1), ("b", 2), ("c", 3))`.
+
+### Pattern semantics
+
+Pack expansions can also appear in patterns. The semantics are chosen to follow
+the general principle that pattern matching is the inverse of expression
+evaluation, so for example if the pattern `(..., [:]x: String)` matches some
+scrutinee value `s`, the expression `(..., [:]x)` should be equal to `s`.
+
+These are run-time semantics, so the scrutinee expression is fully evaluated
+(doesn't contain pack expansions).
+
+**NOTE TO READER:** This section is specifying run-time semantics, but I'm
+paying special attention to the conditions for pattern irrefutability because
+that's going to be an important property when typechecking variadic patterns.
+
+_Tuple pattern:_
+
+There can be no more than one `...,` pattern in a tuple pattern.
+
+The N elements of the pattern before the `...,` expansion are matched with the
+first N elements of the scrutinee, and the M elements of the pattern after the
+`...,` expansion are matched with the last M elements of the scrutinee. If the
+scrutinee does not have at least N + M elements, the pattern does not match. The
+operand of the `...,` pattern is iteratively matched with each of the remaining
+K scrutinee elements, if any. The iterations are numbered sequentially from 1 to
+K, and the number of the current iteration is called the _pack index_.
+
+Observe that the pattern is irrefutable for a given scrutinee type if:
+
+-   At most one `...,` pattern
+-   Scrutinee type is a tuple type with at least N + M elements
+-   First N and last M pattern elements are irrefutable given the corresponding
+    scrutinee element types
+-   For all I from 1 to K, the `...,` operand pattern with pack index I is
+    irrefutable given the type of the Ith scrutinee element. (Note that the
+    typechecker doesn't know K)
```

Comment:
Do we allow anything where there is a single `...,` but it applies to multiple bindings each with a `[:]`
```
fn AddKeyValuePairsToContainer[addr self: Self*](
    ..., [:]keys: [KeyType;], [:]values: [ValueType;]);
// keys and values must have the same arity.
```
Not obvious whether it would mean `(key1, key2, key3, value1, value2, value3)` or `(key1, value1, key2, value2, key3, value3)`, but something like this does seem useful, and still well-defined as long as there are an even number of elements in the scrutinee.

Otherwise, looking at the examples above, I only see these cases:

-   `some_name: a_type(..., [:]some_array)` where `some_name` is going to bind to a single argument, but that argument has a type that varies in a variadic way, like `Apply`.
-   `..., [:]some_name: some_array_of_types`, like `SumInts`, `StrCat`, and `Min`
-   `..., some_name: a_type([:]some_array)`, like `Zip`

It seems better to just list these cases explicitly and say what they mean instead of trying to come up with more general rules.

Further, looking at these last two cases, I'm surprised by how different they look. Generalizing from the `Zip` example, I'd expect the `[:]` to be consistently to the right of the `:`.

Reply:
To be clear, I'm not actually advocating for supporting that first example. While the `(key1, value1, key2, value2, key3, value3)` construct seems interesting, I'd be fine saying that isn't something we are going to support, and the caller will have to pass `((key1, value1), (key2, value2), (key3, value3))` instead. I do think it is a choice we should be making explicitly, though, either favoring expressiveness or reducing the surface area of this feature.